### ISSUE TITLE
Incorrect error in classmethod: Argument 2 "super" not instance of argument 1

### ISSUE THREAD

**Issue by bukzor** [python/mypy#12885]
**Bug Report**

This error was obviously intended for use in a method, but (in a classmethod, especially) argument 2 of super is not required to be an instance of argument 1 -- it's also allowed to be a subclass of argument 1. 

This incorrect error message is masking the actual cause of the failure, making it quite difficult to resolve.

**To Reproduce**

code:

```!python
import typing
from typing import Type

T = typing.TypeVar("T")

del typing  # prevent accidental use below

class MySuper:
  @classmethod
  def f(cls: Type[T]) -> T:
    return cls()

class MySub(MySuper):
  @classmethod
  def f(cls: Type[T]) -> T:
    return super().f()
```

**Expected Behavior**

Successful typechecking.

**Actual Behavior**

```
$ mypy subclassing_classmethod.py
subclassing_classmethod.py:18: error: Argument 2 for "super" not an instance of argument 1
```

**Your Environment**

- Mypy version used: mypy 0.960 (compiled: yes)
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: Python 3.9.12
- Operating system and version: Debian Testing ("Sid") Linux 5.16

**Other Things I Tried**

Other type annotation experiments:

 E0. `T = typing.TypeVar("T")` (above) -- error: Argument 2 for "super" not an instance of argument 1
 E1. `T = typing.TypeVar("T", bound="MySuper")` -- same result
 E2. `T = typing.TypeVar("T", bound="MySuper", covariant=True)`  -- same result
 E3. Separate typevars -- this eliminates the error
    ```!python
    Tsup = typing.TypeVar("Tsup", bound="MySuper")
    Tsub = typing.TypeVar("Tsub", bound="MySub")
    ```
  E4. `T = typing.TypeVar("T", "MySuper", covariant=True)` -- error: only one constraint
  E5. `T = typing.TypeVar("T", "MySuper", "MySuper", covariant=True)` -- succeeds, amusingly
  E6. `T = typing.TypeVar("T", MySuper, MySuper)` -- also succeeds
  E7. `T = typing.TypeVar("T", covariant=True)` -- fails as in E0

In my mind, E2, E4, and E5 should be synonymous in this context. The fact that E5 is valid indicates that perhaps the "more than one constraint" constraint should be lifted. Shouldn't E6 fail?

**Comment by bukzor** [comment]
My questions:

Q1. Why does this error appear in the subclass but not the superclass?
Q2. Is the typesystem preserving the binding from the superclass when analyzing the subclass?
Q3. If so, why doesn't a simple `covariant=True` solve the issue?

**Comment by bukzor** [comment]
pyright rsults for the same eight experiments:

0. T = typing.TypeVar("T") (above) -- Expression of type "MySub" cannot be assigned to return type "T@f"
1. T = typing.TypeVar("T", bound="MySuper") -- *success*
2. T = typing.TypeVar("T", bound="MySuper", covariant=True) -- error: Covariant type variable cannot be used in parameter type 
3. Separate typevars -- this eliminates the error
   ```!python
   Tsup = typing.TypeVar("Tsup", bound="MySuper")
   Tsub = typing.TypeVar("Tsub", bound="MySub")
   ```
4. T = typing.TypeVar("T", "MySuper", covariant=True) -- error: only one constraint, plus "cannot use covariant in parameter"
5. T = typing.TypeVar("T", "MySuper", "MySuper", covariant=True) -- succeeds, amusingly
6. T = typing.TypeVar("T", MySuper, MySuper) -- error: "MySuper" is not defined
6.2. T = typing.TypeVar("T", "MySuper", "MySuper") -- Expression of type "MySuper" cannot be assigned to return type "T@f"
7. T = typing.TypeVar("T", covariant=True) -- fails as in attempt 0, plus "cannot use covariant in parameter"

**Comment by m-aciek** [comment]
For my case there was incorrect bound parameter value set. Providing a hint to the error or changing the error message would be a win for this case.

**Comment by erictraut** [comment]
It appears this bug was fixed in mypy 0.990. I'm not able to repro the issue that version or any subsequent version (including the latest — 1.5).
